docs/architecture/adr-009-modular-antehandler.md

```diff
@@ -157,6 +157,94 @@ Pros:
 Cons:
 1. Cannot wrap antehandlers with decorators like you can with Weave.
 
+### Simple Decorators
+
+This approach takes inspiration from Weave's decorator design while trying to minimize the number of breaking changes to the SDK and maximizing simplicity.
+
+Allow Decorator interface that can be chained together to create an SDK AnteHandler.
+
+This allows users to choose between implementing an AnteHandler by themselves and setting it in the baseapp, or use the decorator pattern to chain their custom decorators with SDK provided decorators in the order they wish.
+
+```golang
+// A Decorator wraps an AnteHandler, and can do pre- and post-processing on the next AnteHandler
+type Decorator interface {
+    AnteHandle(ctx Context, tx Tx, simulate bool, next AnteHandler) (newCtx Context, err error)
+}
+```
+
+```golang
+// ChainDecorators will recursively link all of the Decorators in the chain and return a final AnteHandler function
+// This is done to preserve the ability to set a single AnteHandler function in the baseapp.
+func ChainDecorators(chain ...Decorator) AnteHandler {
```

Comment:
This does decorate the chain of AnteHandlers.
But returns before the Handler is called.

In such a case, the simple ordered-list of ante handlers makes as much sense.
The point of decorators was to wrap the handler execution and be able to do cleanup *after* the *Handler* was called.

If you just want to build a AnteHandler and not call defered code after execution, then the simpler chain approach works well.

Reply:
This decorates the chain of AnteHandlers so that it is possible for one `AnteDecorator` to wrap and cleanup the `next` antehandler. It is clearly more powerful than the chained micro-function approach since the chained micro-function cannot wrap over each other (So it cannot do defer-cleanup).

It's true, this does not allow users to decorate a `MsgHandler` or wrap the entire handler logic with a decorator. It's a deliberate choice to make decorators capable of modifying the authentication and validation work of `AnteHandler`, since that is where most of the use-case for decorators lie in my opinion.

I don't think there's any reason why a decorator for ModuleA should have the power to modify the results from a `MsgHandler` in ModuleB. It does make sense that a decorator in `ModuleA` may perform additional authentication/validation work on a tx, even if it contains a msg for ModuleB.